lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -2050,23 +2050,30 @@ static LogicalResult checkConnectFlow(Operation *connect) {
     // A sink that is a port output or instance input used as a source is okay.
     auto kind = getDeclarationKind(src);
     if (kind != DeclKind::Port && kind != DeclKind::Instance) {
-      auto diag =
-          connect->emitOpError()
-          << "has invalid flow: the right-hand-side has sink flow and "
-             "is not an output port or instance input (expected source "
-             "flow, duplex flow, an output port, or an instance input).";
-      return diag.attachNote(src.getLoc())
-             << "the right-hand-side was defined here.";
+      auto srcRef = getFieldRefFromValue(src);
+      bool rootKnown;
+      auto srcName = getFieldName(srcRef, rootKnown);
+      auto diag = mlir::emitError(connect->getLoc());
+      diag << "connect has invalid flow: the source expression ";
+      if (rootKnown)
+        diag << "\"" << srcName << "\" ";
+      diag << "has sink flow, expected source flow, duplex flow, an output "
+              "port, or an instance input";
```

Comment:
It is probably best to _not_ call out "output port" or "instance input" and just leave it as "expected source or duplex flow".  My reasoning here is that input and output technically have little meaning here other than they factor into flow computation.  E.g., this may give a weird error message stating "b.a [should be] and output port":

```scala
circuit Quz:
  module Quz:
    input a: UInt<1>
    output b: {flip a: UInt<1>}

    b.a <= a
```